### DISCUSSION TITLE
Entity Linker with Transformer Listener

### DISCUSSION THREAD

**Discussion by gblanco10** [explosion/spaCy#12329] [Help: Coding & Implementations] [answered]
I have a pretrained pipeline composed by a transformer and NER components and I am trying to create an Entity Linker able to use embedding representation produced by the transformer rather than using a CNN.
I am adding the Linker to the pipeline as 

```python
entity_linker = nlp.add_pipe("entity_linker",config = {
"labels_discard": [],
   "n_sents": 0,
   "incl_prior": True,
   "incl_context": True,
   "model": {
       "@architectures": "spacy.EntityLinker.v2",
      "tok2vec": {
               "@architectures" : "spacy-transformers.TransformerListener.v1",
               "pooling" : {"@layers":"reduce_mean.v1"},
               "upstream" : "*"
            },
         "nO": 768
      },
   "entity_vector_length": 768,
   "get_candidates": {'@misc': 'spacy.CandidateGenerator.v1'},
   "threshold": None,
})
```

but I receive dimension unset Error

```python
  File "/home/blanco/wiki_kb/.venv/lib/python3.9/site-packages/spacy_transformers/layers/listener.py", line 64, in forward
    width = model.get_dim("nO")
  File "/home/blanco/wiki_kb/.venv/lib/python3.9/site-packages/thinc/model.py", line 175, in get_dim
    raise ValueError(err)
ValueError: Cannot get dimension 'nO' for model 'transformer-listener': value unset
```

Looking online I found that this last problem typically occurs when you try to finetune a late component in the pipeline while freezing the transformer but in my case I am no quite close to that point.

My final aim would be to have a Linker that uses transformer embedding for entity and context and (maybe learning a linear layer for reprojection to a lowest dimension or in case no train at all) compares embedding coming from a knowledge base whose embedding representations have been created using the same pretrained pipeline.

Thanks in advance to everybody who will reply and help me

**Current Environment**
- Python 3.9.13
- spacy 3.4.4
- spacy-alignments 0.9.0
- spacy-legacy 3.0.12
- spacy-loggers 1.0.4
- spacy-transformers 1.2.1

**Comment by adrianeboyd** [comment, marked as answer]
It's hard to know for sure without seeing your full config, but if you want to use a frozen transformer as input to another component, be sure that you're using `spacy` v3.5.0+ and `spacy-transformers` v1.2.2+ due to some bug fixes and be sure that you've added it to `[training.annotating_components]`.

If that doesn't help, can you provide the full config and training steps that you're using?

**Reply by gblanco10** [reply]
Hi @adrianeboyd thanks for your reply.
I have installed `spacy==3.5.0` and `spacy-transformers==1.2.2`, and added `annotating_components` to the config of my pretrained pipeline which now looks like this [config.txt](https://github.com/explosion/spaCy/files/10859881/config.txt) but unfortunately it did not solve the problem.
Regarding the training steps, I was not able to train the linker but I am at the point where I am trying to add it to the pipeline and see its predictions.
This is my main 
```python
import spacy
if spacy.__version__ < "3.5":
    kb_class = spacy.kb.KnowledgeBase
else:
    kb_class = spacy.kb.InMemoryLookupKB

nlp = spacy.load("best_ner_pipeline")
config = {"punct_chars": None}
nlp.add_pipe("sentencizer", config=config)

entity_linker = nlp.add_pipe("entity_linker",config = {
"labels_discard": [],
   "n_sents": 0,
   "incl_prior": True,
   "incl_context": True,
   "model": {
       "@architectures": "spacy.EntityLinker.v2",
        "tok2vec": {
                "@architectures" : "spacy-transformers.TransformerListener.v1",
                "pooling" : {"@layers":"reduce_mean.v1"},
                "upstream" : "*",
                "grad_factor" : 0.0
                },
        "nO": 768
    },
   "entity_vector_length": 768,
   "get_candidates": {'@misc': 'spacy.CandidateGenerator.v1'},
   "threshold": 0.2,
},last=True)

kb = kb_class(vocab=nlp.vocab,entity_vector_length=768)
kb.from_disk(KB_PATH)

entity_linker.initialize(lambda: [example], nlp=nlp, kb_loader=lambda v: kb)
```

and just this last instruction raise an Error about the dimension unset in transformer listener
```python
File "/home/blanco/wiki_kb/.venv/lib/python3.9/site-packages/spacy_transformers/layers/listener.py", line 76, in forward
    width = model.get_dim("nO")
  File "/home/blanco/wiki_kb/.venv/lib/python3.9/site-packages/thinc/model.py", line 175, in get_dim
    raise ValueError(err)
ValueError: Cannot get dimension 'nO' for model 'transformer-listener': value unset
```
Hope this could provide more context in understanding the problem.
Thanks again for your support

**Reply by adrianeboyd** [reply]
I think the problem is that this step isn't going to work on its own. You need to initialize the whole pipeline with the correct `[training]` settings from the config so that the `transformer` has run before `entity_linker` so it can infer the dimensions.

```python
entity_linker.initialize()
```

Basically what happens in `init_nlp` here: https://github.com/explosion/spaCy/blob/74cae47bf65d99dbe50b0fe95f04141779c8005b/spacy/training/initialize.py#L27

I'd strongly recommend training from a config rather a script. It's possible you might run into some rough edges here, but the idea is that you should be able to put your knowledge base paths and settings in the `[initialize]` block. Let me see if I can find an example...

**Reply by adrianeboyd** [reply]
Here's an example from the `tutorials/nel_emerson` project:

https://github.com/explosion/projects/blob/b496a08832dfbdfac716d5a94c815f4c81b18ffe/tutorials/nel_emerson/configs/nel.cfg#L67-L71

**Reply by gblanco10** [reply]
@adrianeboyd thanks again for your support, I will follow your suggestions and your example config changing the `components.entity_linker.model.tok2vec` option and see the outcome. In the meantime, I have just a few comments. 
1. In my example the linker is added as the last component of the pipeline so the transformer has run for sure before it, so why can not it infer the dimension? 
2. At the moment I would like to debug how the linker produces its outcome and be sure it is using transformer embedding. From your comment, I understand I must first produce the final pipeline by training from the config and then load it and debug it, is it true?
3. If I manage to make the linker use transformer embedding it does not have a dedicated machine learning model anymore, so if my intent is to keep the transformer frozen it would not require a training process at all. Would it be possible to add a dedicated linear layer for projecting the embedding (both of input text and from kb candidates) to a smaller dimension? 

Still a huge thanks for your patience and collaboration

**Reply by adrianeboyd** [reply]
If you just call `entity_linker.initialize()`, the `transformer` component isn't run at all so it can't infer the dimensions. You need to initialize the whole pipeline, which is basically what `nlp.initialize()` does, but don't run this exactly because it will wipe out the weights of all the components by default. If you set up the frozen components in your config, `spacy train` will handle the details for you.

The entity linker's internal model currently looks like this:

https://github.com/explosion/spaCy/blob/6aa6b86d496c8d9271f42c077a79f9bfb88687ac/spacy/ml/models/entity_linker.py#L16-L34

There have been a number of recent updates and bug fixes related to frozen transformers so I wanted to make sure there was nothing going wrong on the technical side in the background, so I sketched how you could modify the `tutorials/nel_emerson` project to use transformer embeddings instead of static word vectors with a frozen transformer component. It doesn't work very well (neither using a frozen transformer nor using these particular transformer embeddings would be expected to perform well for this task), but it trains and runs:

https://github.com/adrianeboyd/projects/tree/sketch/entity-linker-frozen-transformer/tutorials/nel_emerson

**Reply by gblanco10** [reply]
Hi @adrianeboyd enormous thanks for your collaboration and for creating a dedicated version of the Emerson project.
I have created my own config [linker_debug.txt](https://github.com/explosion/spaCy/files/10932908/linker_debug.txt) and tried to train the pipeline on a sample corpus for a few epochs to ensure everything works as expected. Unfortunately, I encountered a few issues:
- I have specified nO in `entity_linker.tok2vec` to 128 as I would like to reproject the embeddings in KB and the one extracted from input text (which are both 768) but spacy CLI gives an error
```python
Traceback (most recent call last):
  File "/usr/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/blanco/wiki_kb/.venv/lib/python3.9/site-packages/spacy/__main__.py", line 4, in <module>
    setup_cli()
  File "/home/blanco/wiki_kb/.venv/lib/python3.9/site-packages/spacy/cli/_util.py", line 74, in setup_cli
    command(prog_name=COMMAND)
  File "/home/blanco/wiki_kb/.venv/lib/python3.9/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/home/blanco/wiki_kb/.venv/lib/python3.9/site-packages/typer/core.py", line 778, in main
    return _main(
  File "/home/blanco/wiki_kb/.venv/lib/python3.9/site-packages/typer/core.py", line 216, in _main
    rv = self.invoke(ctx)
  File "/home/blanco/wiki_kb/.venv/lib/python3.9/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/blanco/wiki_kb/.venv/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/blanco/wiki_kb/.venv/lib/python3.9/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/home/blanco/wiki_kb/.venv/lib/python3.9/site-packages/typer/main.py", line 683, in wrapper
    return callback(**use_params)  # type: ignore
  File "/home/blanco/wiki_kb/.venv/lib/python3.9/site-packages/spacy/cli/train.py", line 45, in train_cli
    train(config_path, output_path, use_gpu=use_gpu, overrides=overrides)
  File "/home/blanco/wiki_kb/.venv/lib/python3.9/site-packages/spacy/cli/train.py", line 75, in train
    train_nlp(nlp, output_path, use_gpu=use_gpu, stdout=sys.stdout, stderr=sys.stderr)
  File "/home/blanco/wiki_kb/.venv/lib/python3.9/site-packages/spacy/training/loop.py", line 124, in train
    raise e
  File "/home/blanco/wiki_kb/.venv/lib/python3.9/site-packages/spacy/training/loop.py", line 107, in train
    for batch, info, is_best_checkpoint in training_step_iterator:
  File "/home/blanco/wiki_kb/.venv/lib/python3.9/site-packages/spacy/training/loop.py", line 209, in train_while_improving
    nlp.update(
  File "/home/blanco/wiki_kb/.venv/lib/python3.9/site-packages/spacy/language.py", line 1155, in update
    proc.update(examples, sgd=None, losses=losses, **component_cfg[name])  # type: ignore
  File "/home/blanco/wiki_kb/.venv/lib/python3.9/site-packages/spacy/pipeline/entity_linker.py", line 377, in update
    loss, d_scores = self.get_loss(
  File "/home/blanco/wiki_kb/.venv/lib/python3.9/site-packages/spacy/pipeline/entity_linker.py", line 415, in get_loss
    raise RuntimeError(err)
RuntimeError: [E147] Unexpected error in the get_loss functionality of the EntityLinker: gold entities do not match up. This is likely a bug in spaCy, so feel free to open an issue: https://github.com/explosion/spaCy/issues
```
- I then tried to set nO to null (or 768, the final outcome is the same) and the training process was completed without problems. So I created a scratch code in which I load the trained pipeline and use it to predict an input text. I placed a break point in `Entity Linker predict ` function but I noticed that after model predict
```python
if self.incl_context:
  # get n_neighbour sentences, clipped to the length of the document
  start_sentence = max(0, sent_index - self.n_sents)
  end_sentence = min(
      len(sentences) - 1, sent_index + self.n_sents
  )
  start_token = sentences[start_sentence].start
  end_token = sentences[end_sentence].end
  sent_doc = doc[start_token:end_token].as_doc()
  # currently, the context is the same for each entity in a sentence (should be refined)
  sentence_encoding = self.model.predict([sent_doc])[0]
```
the variable `sentence_encoding` is a tensor full of 0 values, even with different input texts. I checked the second layer in `entity_linker.model.layers` and it is a linear layer with dimensions (768,768) so it just reprojects input text in the same dimension which is fine given the input config, but it is strange to obtain a zeros array. Looking at the linker source code I noticed also that the linear layer is not applied to KB embedding, but just to input text so I understand I am not allowed to project both text and entities to a lower dimension, which may be the reason behind the error I got when setting nO to 128.

Looking forward to hearing a comment from you about this.

Thanks again for your collaboration

**Reply by adrianeboyd** [reply]
Oh, sorry, I was testing with `master` rather than v3.5.0 and there are some KB fixes that are coming out in v3.5.1 very soon.

**Reply by adrianeboyd** [reply]
I think one problem is that `Span.as_doc()` is not going to copy or slice the data in `Doc._.trf_data` that is needed for the transformer embedding vector hook (and this would be non-trivial due to the overlapping strided spans). There could be some options related to manually setting `Span.vector`, but then I think you still have the issue that the existing vector hooks would still override it, so it looks like there are a lot of internal details that just haven't been worked out for transformers yet.

We are currently working on a lot of entity linker improvements, let me hand this off to @rmitsch who knows more about the details and plans.

**Reply by gblanco10** [reply]
I understand that it might be extremely complicated adapting this complex code base so thanks again for the collaboration and I am looking forward to next updates.

In the meantime I tried to work on a custom version of the Entity Linker by overriding the base class.

I have thought that entity context in input text and entity description in KB might be different so it could make sense having two distinct linear architectures, one for projecting input text embedding and one for projecting KB embedding, so I came up with this Entity Linker architecture and factory class [trf_linker.txt](https://github.com/explosion/spaCy/files/10944248/trf_linker.txt) that I am trying to train with this [trf_linker_debug.txt](https://github.com/explosion/spaCy/files/10944254/trf_linker_debug.txt).

Do you think it could make sense ? According to first little experiments it is able to run without problems, kind harder understanding whether it behaves as expected

Thanks again for your help

**Reply by rmitsch** [reply]
Hi @gblanco10! I think you're on the right track. We are aware that having static KB embeddings - i.e. not tuning them w.r.t. the objective of them being similar with the embeddings of relevant input text - is not ideal. We are hence currently working on modifying our NEL architecture so that both KB and text input embeddings are fine-tuned to this end.

As for your custom `EntityLinker` implementation - without having reviewed it in detail, I think it sounds reasonable. I'd encourage you to benchmark your solution on some standard datasets to figure out whether it behaves as expected (we'd also be interested in your results!).

**Reply by gblanco10** [reply]
Hi, @rmitsch, thanks for your reply!

Having static KB embedding is not a problem as soon as the transformer is frozen, which is what happens in my situation. Still, I understand that in those situations where transformer weights are constantly updated it makes sense to update KB representations accordingly. 

Anyway, I will test my solution with some benchmarks, thanks for the encouragement! To this end, do you have any suggestions about the datasets to use? In case are there spacy-compatible corpus versions of these datasets? I have thought about [AIDA CoNLL-YAGO](https://paperswithcode.com/dataset/aida-conll-yago), but I am looking forward to other suggestions.

Thanks in advance for your support

**Reply by rmitsch** [reply]
If the license works for you, AIDA CoNLL-YAGO is a decent choice. I can also recommend the [Mewsli-9 dataset](https://github.com/google-research/google-research/tree/master/dense_representations_for_entity_retrieval/mel#get-mewsli-9-dataset) ([paper](https://aclanthology.org/2020.emnlp-main.630.pdf)).

If you are looking for inspiration, our current NEL benchmark (WIP, currently running on just Mewsli-9) is available [here](https://github.com/explosion/projects/tree/v3/benchmarks/nel).

**Reply by gblanco10** [reply]
Hi, @rmitsch I have tried to run the pipeline with Mewsli-9 dataset following spacy NEL benchmark you listed.
I have run all the preliminary steps already defined in project template for data download and preprocessing and I have slightly modified the script for KB creation in order to use transformer pipeline.

Then I have trained both baseline configuration [baseline_nel.txt](https://github.com/explosion/spaCy/files/11003318/baseline_nel.txt) and one with my entity linker transformer [trf_nel.txt](https://github.com/explosion/spaCy/files/11003319/trf_nel.txt).


According to training logs ( [baseline_logs.csv](https://github.com/explosion/spaCy/files/11003344/baseline_logs.csv) [trf64_logs.csv](https://github.com/explosion/spaCy/files/11003347/trf64_logs.csv) ) both pipelines reach 100 of nel micro_f1 which sounds suspiciously good, in fact when I try to evaluate both pipelines (always using the provided scripts) I get strange results with 0s in all rows except for candidate generation 
[evaluation_both.csv](https://github.com/explosion/spaCy/files/11003394/evaluation_both.csv)

I am not sure about what I did wrong in the process, do you have any suggestion ? 

Thanks again for your support.

**Reply by rmitsch** [reply]
To be clear, you ran the NEL benchmark project with the default configuration (without any further changes) and got a `micro_f1` of 100 and the 0 scores as included in `evaluation_both.csv`?

**Reply by gblanco10** [reply]
Yes exactly, I have trained the attached configurations and got the corresponding training logs. While on the test set I got the exact same result for both of them

**Reply by rmitsch** [reply]
Thanks for reporting back. We will take a look at this.

**Reply by gblanco10** [reply]
Hi @rmitsch, have you had a chance to look at the experiment results? Thanks

**Reply by rmitsch** [reply]
Sorry for my late response! No, not yet. It's on my to-do list, but I haven't gotten around to it yet. I'll update here :)

**Reply by rmitsch** [reply]
Short update - I haven't gotten around to this yet, but I should be able to do so soon. Will keep you in the loop!

**Reply by rmitsch** [reply]
My apologies for the late response! So, I finally ran the NEL benchmark with the current `main` of https://github.com/explosion/projects. I ran every single step without modifications and ended up with reasonable, non-zero scores. Are you sure you ran the vanilla configuration on a fresh venv?

**Comment by jkgenser** [comment]
> We are hence currently working on modifying our NEL architecture so that both KB and text input embeddings are fine-tuned to this end.

@rmitsch This sounds very interesting. I've had success with current NEL model and interested in whether you have documented intended plans here or literature or other resources the new work is inspired by?

**Reply by rmitsch** [reply]
Happy to hear that the current NEL model worked for you! We have no publicly available documentation on the intended plans. What I can tell you at this point is that we are thinking of experimenting with the DEER architecture as introduced [here](https://arxiv.org/abs/1909.10506).

**Reply by jkgenser** [reply]
Will look into that.  For my use case, I've used a fuzzy search candidate generator because there is a large space of potential aliases and most are not present in training data. i.e. "a. fib, "a. fib.", etc. which map to "atrial fibrilation" This approach has worked well but I've also been debating using ANN as part of candidate generation as well as debating removing the linker component entirely and fine-tuning a transformer pipeline on NER, then taking the finetuned transformer and then encoding mention+context as well as kb entity + description and doing ANN as the entity resolution step itself; kind of like sciscpacy did but instead of using tfidf 3/4 chargrams using the transformer encoding. I suppose CNN encodings would work similarly. 

One of the biggest issues I find is that that I have the encode the entity descriptions into the knowledge base **before** fine-tuning on NER+NEL task which is updating weights back into my transformer (or CNN I suppose) and therefore learning a better encoding of mentions + descriptions.

Do you think the approach I describe is reasonable stopgap before the DEER approach is implemented into spacy? (FYI: @svlandeg also since I know you're another spacy expert on KB things - no pressure to weigh , tagging you in case hearing from another user and use case is useful)

That paper looks interesting because it would involve learning representations for an NN search which also seems to be the direction a lot of the literature is going. Looking forward to it!

**Reply by rmitsch** [reply]
> One of the biggest issues I find is that that I have the encode the entity descriptions into the knowledge base before fine-tuning on NER+NEL task which is updating weights back into my transformer (or CNN I suppose) and therefore learning a better encoding of mentions + descriptions.

IMO one of the advantages of DEER is that the "fine-tuning" happens via a separate model on top of the models you use to infer your base embeddings. I. e. it's not really fine-tuning in that sense, but mapping the embeddings for entity descriptions and textual mentions to another embedding space, with the objective of maximizing the similarity of the resulting embeddings of the description of entity `E` and the mention `M` if `M` should resolve to `E`. So you have a relatively light-weight model to adjust the embeddings you're getting out of your transformer model, and it's learning to do so for entities and mentions in tandem. At inference time an ANN vector search operation is sufficient to retrieve the most relevant candidates (it might be beneficial to add another linker model on top, we'll test that).

> Do you think the approach I describe is reasonable stopgap before the DEER approach is implemented into spacy?

It sounds reasonable to me. If you do that, we'd be happy to hear about your results!
